File: go/classic/minStack_30.go
package classic

import (
	"container/list"
)

type MinStack struct {
	data *list.List
	min  *list.List
}

/** initialize your data structure here. */
func ConstructorForMinStack() MinStack {
	return MinStack{
		data: list.New(),
		min:  list.New(),
	}
}

func (this *MinStack) Push(x int) {
	if this.min.Len() == 0 {
		this.min.PushBack(x)
		this.data.PushBack(x)
		return
	}
	if x < this.min.Back().Value.(int) {
		this.min.PushBack(x)
	} else {
		this.min.PushBack(this.min.Back().Value)
	}
	this.data.PushBack(x)
}

func (this *MinStack) Pop() {
	if this.data.Len() > 0 {
		this.data.Remove(this.data.Back())
		this.min.Remove(this.min.Back())
	}
}

func (this *MinStack) Top() int {
	if this.data.Len() > 0 {
		return this.data.Back().Value.(int)
	}
	return -1
}

func (this *MinStack) Min() int {
	if this.min.Len() > 0 {
		return this.min.Back().Value.(int)
	}
	return -1
}

File: java/jzoffer/MinStack_30.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class MinStack_30 {

    LinkedList<Integer> min = new LinkedList<>();
    LinkedList<Integer> nums = new LinkedList<>();
    /** initialize your data structure here. */
    public MinStack_30() {}

    public void push(int x) {
        nums.add(x);
        if (min.isEmpty()) {
            min.add(x);
            return;
        }
        min.add(Math.min(x, min.get(min.size() - 1)));
    }

    public void pop() {
        min.removeLast();
        nums.removeLast();
    }

    public int top() {
        return nums.getLast();
    }

    public int min() {
        return min.getLast();
    }
}
